elements/lisk-p2p/src/peer_book/base_list.ts

```diff
@@ -13,146 +13,192 @@
+			return undefined;
+		}
+
+		return sanitizeInternalPeerInfo(peerInfo);
 	}
 
-	public addPeer(peerInfo: P2PPeerInfo): CustomPeerInfo | undefined {
-		if (this.getPeer(peerInfo)) {
-			throw new ExistingPeerError(peerInfo);
+	public addPeer(
+		incomingPeerInfo: P2PEnhancedPeerInfo,
+	): P2PEnhancedPeerInfo | undefined {
+		const existingPeerInfo = this.getPeer(incomingPeerInfo.peerId);
+		if (existingPeerInfo) {
+			throw new ExistingPeerError(incomingPeerInfo);
 		}
-		const bucket = this.getBucket(peerInfo.ipAddress);
-		const incomingPeerId = peerInfo.peerId;
-		const newPeer = this.initPeerInfo(peerInfo);
-		const evictedPeer = this.makeSpace(peerInfo.ipAddress);
-		bucket.set(incomingPeerId, newPeer);
 
-		// If a peer was evicted in order to make space for the new one, we return its info
+		const { bucketId, bucket } = this.calculateBucket(
+			incomingPeerInfo.ipAddress,
+			this.type === PEER_TYPE.NEW_PEER
+				? incomingPeerInfo.sourceAddress
+				: undefined,
+		);
+
+		// If bucket is full, evict a peer to make space for incoming peer
+		const evictedPeer =
+			bucket.size === this.peerListConfig.bucketSize
+				? this.makeSpace(bucket)
+				: undefined;
+
+		const internalPeerInfo = {
+			...incomingPeerInfo,
+			numOfConnectionFailures: 0,
+			dateAdded: new Date(),
+			bucketId,
+		};
+		bucket.set(incomingPeerInfo.peerId, internalPeerInfo);
+		this.peerIdToPeerInfo.set(incomingPeerInfo.peerId, internalPeerInfo);
+
 		return evictedPeer;
 	}
 
-	public updatePeer(peerInfo: P2PPeerInfo): boolean {
-		const bucket = this.getBucket(peerInfo.ipAddress);
-		const incomingPeerId = peerInfo.peerId;
-		const foundPeer = bucket.get(incomingPeerId);
+	public updatePeer(incomingPeerInfo: P2PPeerInfo): boolean {
+		const bucket = this.getBucket(incomingPeerInfo.peerId);
 
-		if (!foundPeer) {
+		if (!bucket) {
 			return false;
 		}
-		const updatedPeerInfo: CustomPeerInfo = {
-			...foundPeer,
-			peerInfo: { ...foundPeer.peerInfo, ...peerInfo },
+
+		const updatedInternalPeerInfo = {
+			...bucket.get(incomingPeerInfo.peerId),
+			...incomingPeerInfo,
 		};
 
-		bucket.set(incomingPeerId, updatedPeerInfo);
+		bucket.set(incomingPeerInfo.peerId, updatedInternalPeerInfo);
+		this.peerIdToPeerInfo.set(incomingPeerInfo.peerId, updatedInternalPeerInfo);
 
 		return true;
 	}
 
-	public removePeer(peerInfo: P2PPeerInfo): boolean {
-		const bucket = this.getBucket(peerInfo.ipAddress);
-		const incomingPeerId = peerInfo.peerId;
+	public removePeer(incomingPeerInfo: P2PPeerInfo): boolean {
+		const bucket = this.getBucket(incomingPeerInfo.peerId);
 
-		if (bucket.get(incomingPeerId)) {
-			const result = bucket.delete(incomingPeerId);
+		if (bucket && bucket.get(incomingPeerInfo.peerId)) {
+			const removedFromBucket = bucket.delete(incomingPeerInfo.peerId);
+			const removedFromPeerLookup = this.peerIdToPeerInfo.delete(
+				incomingPeerInfo.peerId,
+			);
 
-			return result;
+			return removedFromBucket && removedFromPeerLookup;
```

Comment:
shouldn't this be `||` ? although as long as it's consistent, it shouldn't matter tho

Reply:
I think it should be `&&` to enforce consistency. Technically it should never happen but maybe strict is better?